subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/CollectingMacroLookup.java

```diff
@@ -38,14 +41,24 @@ public void append(File file, IncludeDirectives includeDirectives) {
             // Ignore
             return;
         }
+
         if (visible == null) {
             visible = new LinkedHashMap<File, IncludeDirectives>();
             visible.put(file, includeDirectives);
+
+            hash = Objects.hashCode(hash, includeDirectives.getMacros().values(), includeDirectives.getMacrosFunctions().values());
```

Comment:
We shouldn't use `Object.hashCode()` (which is what this is doing), as the result will have too many collisions due to weak contracts. To calculate a fingerprint for an object graph, we should use something like `DefaultBuildCacheHasher` to calculate a hash, or serialize the graph using a `Serializer` and run the result through a message digest (both options are basically doing the same thing). Using a `Serializer` would be a better option here as we already know how to serialise an `IncludeDirectives` to a stream.